### ISSUE TITLE
Performance of Promotion applying with many promotions

### ISSUE THREAD

**Issue by stefandoorn** [Sylius/Sylius#9882]
**Sylius version affected**: 1.1.x

E.g. in file `Sylius/Component/Promotion/Processor/PromotionProcessor.php` this call is made:

`$preQualifiedPromotions = $this->preQualifiedPromotionsProvider->getPromotions($subject);`

This basically fetches all the active promotions available for a single order (`$subject`) based on channel.

After that, it loops over each to see whether they are valid:

```
 foreach ($preQualifiedPromotions as $promotion) {
            if (!$promotion->isExclusive() && $this->promotionEligibilityChecker->isEligible($subject, $promotion)) {
                $this->promotionApplicator->apply($subject, $promotion);
            }
        }
```

In this case, probably less promotions have to be fetched from the database and checked against their rules:

* The ones with `exclusive = 1` can be dropped already (this is the easiest..) :-)
* Plus, some smart rules around being eligible would also be possible here. This is obviously the hard part, but e.g. based on the order, some rules are possibly checked already:
** cart_quantity: we know the amount in the order
** has_taxon: we know the products, therefore the taxon
** total_of_items_from_taxon: same here
** contains_product: we know the products in the order
** item_total: we know it from the order

The one with customer group is harder. Why I'm coming up with this, is that in our current shop we calculate product discounts throughout the frontend, so it's no in the cart/checkout yet (as Sylius only does by default). Therefore, showing lots of products on a page does a lot of promotions calculations. I've implemented some caching, but if it can be solved in the root of this processing part, I think it will even be more efficient and beneficial for cart & checkout. Let's say you have 1000 promotions, it might potentially need to loop over all 1000 before it calculated the discount for this order. That's clearly an 1n issue there.

It's not something easily solved, I know, but it might be worth trying.

**Comment by stefandoorn** [comment]
P.S. Using that method with 18 promotions in the database with each 1 promotion rule connected to them, results in 18 times a DB query to fetch the rule. Maybe that can be fixed by itself, loading the promotion always including the rules directly (some hydrator / lazy loading?).

**Comment by stefandoorn** [comment]
Quick PoC: putting a hydrator on `findActiveByChannel` in the promotion repository makes my dev environment go from 18 to 1 query to fetch the rules for 18 promotions. I think it's good to stick to just the rules, as we only need the actions as soon as we know they have to be applied.

**Comment by pamil** [comment]
Fixed by #9883.

**Comment by stefandoorn** [comment]
Well, I think there is still room for thought on this but that's for 2.0: whether finding the applicable rules can be done via SQL instead of inside PHP. But that probably needs quite some rewriting.
